contracts/mapping-token/contracts/darwinia/DarwiniaMappingTokenFactory.sol

```diff
@@ -109,16 +120,51 @@ contract DarwiniaMappingTokenFactory is Initializable, Ownable {
         IERC20(token).mint(recipient, amount);
     }
+    
+    // cross transfer to remote chain without waiting any confirm information,
+    // this must request the burn proof can be always verified by the remote chain corrently.
+    // so, here the user's token burned directly.
+    function burnAndRemoteUnlock(uint32 specVersion, uint64 weight, address token, bytes memory recipient, uint256 amount) external payable {
+        burnAndSendProof(specVersion, weight, token, recipient, amount);
+        IERC20(token).burn(address(this), amount);
+    }
+
+    // cross transfer to remote chain with waiting confirmation.
+    // the transfered token information locked in the contract first, then wait the result from remote chain.
+    function burnAndRemoteUnlockWaitingConfirm(uint32 specVersion, uint64 weight, address token, bytes memory recipient, uint256 amount) external payable {
+        burnAndSendProof(specVersion, weight, token, recipient, amount);
+        TokenInfo memory info = tokenToInfo[token];
+        (bool readSuccess, bytes memory messageId) = DISPATCH_ENCODER.call(
+            abi.encodePacked(info.eventReceiver, bytes4(keccak256("read_latest_message_id()")))
+        );
+        require(readSuccess, "burn: read message id failed");
+        transferUnconfirmed[messageId] = UnconfirmedInfo(msg.sender, token, amount);
+        emit BurnAndWaitingConfirm(messageId, msg.sender, recipient, token, amount);
+    }
+
+    // confirm transfer information from remote chain.
+    // 1. if event is verified and the token unlocked successfully on remote chain, then we burn the mapped token
+    // 2. if event is verified, but the token can't unlocked on remote chain, then we take back the mapped token to user.
+    function confirmBurnAndRemoteUnlock(bytes memory messageId, bool result) external onlySystem {
+        UnconfirmedInfo memory info = transferUnconfirmed[messageId];
+        require(info.amount > 0 && info.sender != address(0) && info.token != address(0), "invalid unconfirmed message");
+        if (result) {
+            IERC20(info.token).burn(address(this), info.amount);
+        } else {
+            require(IERC20(info.token).transfer(info.sender, info.amount), "transfer back failed");
+        }
+        delete transferUnconfirmed[messageId];
+        emit RemoteUnlockConfirmed(messageId, info.sender, info.token, info.amount, result);
+    }
 
-    function crossTransfer(uint32 specVersion, uint64 weight, address token, bytes memory recipient, uint256 amount) external payable {
+    function burnAndSendProof(uint32 specVersion, uint64 weight, address token, bytes memory recipient, uint256 amount) internal {
         require(amount > 0, "can not transfer amount zero");
         TokenInfo memory info = tokenToInfo[token];
         require(info.source != address(0), "token is not created by factory");
         require(IERC20(token).transferFrom(msg.sender, address(this), amount), "transfer token failed");
```

Comment:
Add comment,

Lock the fund in this before message on remote backing chain get dispatched successfully and burn finally